File: crates/blazen-node/examples/branching_workflow.mjs
/**
 * Branching (Fan-Out) Workflow Example
 *
 * Demonstrates conditional branching in a Blazen workflow. A single step can
 * return an ARRAY of event objects, causing the workflow to "fan out" into
 * multiple parallel branches. Each downstream step declares the event types
 * it listens for, so only the matching branch receives each event.
 *
 * Because multiple branches may race toward completion, the first
 * blazen::StopEvent to fire wins and the workflow resolves with that result.
 *
 * Scenario:
 *   1. The "analyze" step inspects a piece of text for sentiment.
 *   2. If the text is positive it emits a "Positive" event; if negative, a
 *      "Negative" event. (Only one event is emitted -- this is conditional
 *      routing, not a true parallel fan-out, but the mechanism is the same.)
 *   3. Dedicated handler steps for each sentiment produce the final result.
 *
 * Run with: node branching_workflow.mjs
 */

import { Workflow } from "blazen";

// --- Build the workflow --------------------------------------------------- //

const wf = new Workflow("sentiment-branching");

/**
 * Step 1 -- Analyze sentiment.
 *
 * Listens for the built-in StartEvent. Performs a naive keyword check and
 * returns an array of events. Returning an array is what triggers fan-out;
 * here we return a single-element array for conditional routing, but you
 * could return multiple elements to fan out into several branches at once.
 */
wf.addStep("analyze", ["blazen::StartEvent"], async (event, ctx) => {
  const text = event.data?.text ?? "";
  console.log(`[analyze] Received text: "${text}"`);

  // Naive sentiment check -- look for a handful of positive keywords.
  const positiveWords = ["great", "awesome", "love", "happy", "fantastic", "good", "wonderful"];
  const isPositive = positiveWords.some((w) => text.toLowerCase().includes(w));

  if (isPositive) {
    console.log("[analyze] Detected POSITIVE sentiment -> routing to handle_positive");
    return [{ type: "Positive", text }];
  }

  console.log("[analyze] Detected NEGATIVE sentiment -> routing to handle_negative");
  return [{ type: "Negative", text }];
});

/**
 * Step 2a -- Handle positive sentiment.
 *
 * Only fires when a "Positive" event is emitted.
 */
wf.addStep("handle_positive", ["Positive"], async (event, ctx) => {
  console.log("[handle_positive] Processing positive text");
  return {
    type: "blazen::StopEvent",
    result: { sentiment: "positive", reply: "Glad to hear it!" },
  };
});

/**
 * Step 2b -- Handle negative sentiment.
 *
 * Only fires when a "Negative" event is emitted.
 */
wf.addStep("handle_negative", ["Negative"], async (event, ctx) => {
  console.log("[handle_negative] Processing negative text");
  return {
    type: "blazen::StopEvent",
    result: { sentiment: "negative", reply: "Sorry to hear that." },
  };
});

// --- Run the workflow ----------------------------------------------------- //

const input = { text: "I had a fantastic day at the park!" };

console.log("--- Running sentiment-branching workflow ---");
console.log();

const result = await wf.run(input);

console.log();
console.log(`Branch taken : ${result.data.sentiment}`);
console.log(`Reply        : ${result.data.reply}`);

File: crates/blazen-py/examples/branching_workflow.py
"""Conditional branching (fan-out) in a Blazen workflow.

Demonstrates how a single step can dispatch multiple events to different
downstream steps, creating parallel branches that race to completion.

Key concepts:
  - **Fan-out**: Return a list of Events from a step to dispatch them all
    at once.  Each event is routed independently based on its event type.
  - **Routing**: Use ``accepts=["EventType"]`` on a step to subscribe it
    to a specific event type.  Steps without an explicit ``accepts`` only
    receive ``StartEvent``.
  - **First StopEvent wins**: When several branches race, the first
    ``StopEvent`` to arrive terminates the workflow and becomes the result.

Run with: python branching_workflow.py
"""

import asyncio

from blazen import Context, Event, StartEvent, StopEvent, Workflow, step

# A handful of positive and negative keywords used for our toy classifier.
POSITIVE_WORDS = {"great", "wonderful", "fantastic", "excellent", "happy", "love", "amazing"}
NEGATIVE_WORDS = {"terrible", "awful", "horrible", "bad", "hate", "worst", "sad"}


# ---------------------------------------------------------------------------
# Steps
# ---------------------------------------------------------------------------


@step  # Accepts StartEvent by default
async def analyze_sentiment(ctx: Context, ev: Event):
    """Examine the input text and route to the appropriate sentiment branch.

    Returns a *list* of one Event -- either PositiveSentiment or
    NegativeSentiment -- so the downstream handler for that type picks it up.
    Returning a list (even of length one) is the idiomatic fan-out pattern;
    you could return two or more events to trigger multiple branches in
    parallel.
    """
    text: str = ev.text.lower()

    # Count keyword hits for each sentiment.
    pos_hits = sum(1 for w in POSITIVE_WORDS if w in text)
    neg_hits = sum(1 for w in NEGATIVE_WORDS if w in text)

    # Store the original text in shared context so downstream steps can read it.
    ctx.set("original_text", ev.text)

    if pos_hits >= neg_hits:
        # Route to the positive branch.
        return [Event("PositiveSentiment", score=pos_hits)]
    else:
        # Route to the negative branch.
        return [Event("NegativeSentiment", score=neg_hits)]


@step(accepts=["PositiveSentiment"])
async def handle_positive(ctx: Context, ev: Event):
    """Handle text that was classified as positive."""
    original = ctx.get("original_text")
    return StopEvent(result={
        "sentiment": "positive",
        "confidence_hits": ev.score,
        "text": original,
    })


@step(accepts=["NegativeSentiment"])
async def handle_negative(ctx: Context, ev: Event):
    """Handle text that was classified as negative."""
    original = ctx.get("original_text")
    return StopEvent(result={
        "sentiment": "negative",
        "confidence_hits": ev.score,
        "text": original,
    })


# ---------------------------------------------------------------------------
# Main
# ---------------------------------------------------------------------------


async def main():
    # Build the workflow from all three steps.
    wf = Workflow("sentiment-branching", [analyze_sentiment, handle_positive, handle_negative])

    samples = [
        "I had a wonderful and fantastic day at the park!",
        "That was the worst experience, absolutely terrible service.",
        "The weather is okay today, nothing special.",
    ]

    for text in samples:
        # Kick off the workflow; keyword arguments become StartEvent attributes.
        handler = await wf.run(text=text)
        result = await handler.result()

        data = result.to_dict()
        branch = data.get("sentiment", "unknown")
        hits = data.get("confidence_hits", 0)

        print(f"Text:      {text}")
        print(f"Branch:    {branch}")
        print(f"Keyword hits: {hits}")
        print()


if __name__ == "__main__":
    asyncio.run(main())
